lib/OperatingSystems/Linux.php

```diff
@@ -0,0 +1,281 @@
+ *
+ */
+
+namespace OCA\ServerInfo\OperatingSystems;
+
+use OCA\ServerInfo\Resources\Disk;
+use OCA\ServerInfo\Resources\Memory;
+use OCA\ServerInfo\Resources\NetInterface;
+use RuntimeException;
+
+class Linux implements IOperatingSystem {
+	private const AF_INET = 2;
+	private const AF_INET6 = 10;
+
+	public function supported(): bool {
+		return true;
+	}
+
+	public function getMemory(): Memory {
+		$data = new Memory();
+
+		try {
+			$meminfo = $this->readContent('/proc/meminfo');
+		} catch (RuntimeException $e) {
+			return $data;
+		}
+
+		$matches = [];
+		$pattern = '/(?<Key>(?:MemTotal|MemFree|MemAvailable|SwapTotal|SwapFree)+):\s+(?<Value>\d+)\s+(?<Unit>\w{2})/';
+
+		$result = preg_match_all($pattern, $meminfo, $matches);
+		if ($result === 0 || $result === false) {
+			return $data;
+		}
+
+		foreach ($matches['Key'] as $i => $key) {
+			// Value is always in KB: https://github.com/torvalds/linux/blob/c70672d8d316ebd46ea447effadfe57ab7a30a50/fs/proc/meminfo.c#L58-L60
+			$value = (int)((int)$matches['Value'][$i] / 1024);
+
+			switch ($key) {
+				case 'MemTotal':
+					$data->setMemTotal($value);
+					break;
+				case 'MemFree':
+					$data->setMemFree($value);
+					break;
+				case 'MemAvailable':
+					$data->setMemAvailable($value);
+					break;
+				case 'SwapTotal':
+					$data->setSwapTotal($value);
+					break;
+				case 'SwapFree':
+					$data->setSwapFree($value);
+					break;
+			}
+		}
+
+		return $data;
+	}
+
+	public function getCpuName(): string {
+		$data = 'Unknown Processor';
+
+		try {
+			$cpuinfo = $this->readContent('/proc/cpuinfo');
+		} catch (RuntimeException $e) {
+			return $data;
+		}
+
+		$matches = [];
+
+		if (str_contains($cpuinfo, 'Raspberry Pi')) {
+			$pattern = '/Model\s+:\s(.+)/';
+		} elseif (str_contains($cpuinfo, 'PowerNV') || str_contains($cpuinfo, 'CHRP IBM pSeries')) {
+			$pattern = '/cpu\s+:\s+(.+)/';
+		} else {
+			$pattern = '/model name\s:\s(.+)/';
+		}
+
+		$result = preg_match_all($pattern, $cpuinfo, $matches);
+		if ($result === 0 || $result === false) {
+			return $data;
+		}
+
+		$model = $matches[1][0];
+
+		$pattern = '/processor\s+:\s(.+)/';
+
+		preg_match_all($pattern, $cpuinfo, $matches);
+		$cores = count($matches[1]);
+
+		if ($cores === 1) {
+			$data = $model . ' (1 core)';
+		} else {
+			$data = $model . ' (' . $cores . ' cores)';
+		}
+
+		return $data;
+	}
+
+	public function getTime(): string {
+		return (string)shell_exec('date');
+	}
+
+	public function getUptime(): int {
+		$data = -1;
+
+		try {
+			$uptime = $this->readContent('/proc/uptime');
+		} catch (RuntimeException $e) {
+			return $data;
+		}
+
+		[$uptimeInSeconds,] = array_map('intval', explode(' ', $uptime));
+
+		return $uptimeInSeconds;
+	}
+
+	public function getNetworkInfo(): array {
+		$result = [];
+		$result['hostname'] = \gethostname();
+		$dns = shell_exec('cat /etc/resolv.conf |grep -i \'^nameserver\'|head -n1|cut -d \' \' -f2');
```

Comment:
`grep` originally parses files, and it can stop at the first match:
```suggestion
		$dns = shell_exec('grep -im1 \'^nameserver\' /etc/resolv.conf |cut -d \' \' -f2');
```
Or with a single executable call, assuming that `awk` is available like `grep` on every Linux system:
```suggestion
		$dns = shell_exec('awk \'/^nameserver/{print $2;exit}\' /etc/resolv.conf');
```
I see `awk` is used below as well, so this would be the preferable option to minimise system load and maximise performance. But I guess that should better go into a dedicated PR. I can open one in case and add some other similar little enhancements.

Reply:
Thanks, sounds like a good plan.